crates/ide/src/lib.rs

```diff
@@ -119,6 +119,12 @@ impl<T> RangeInfo<T> {
     }
 }
 
+#[derive(Debug)]
+pub struct Workspace {
+    pub root_dir: PathBuf,
+    pub target_dir: PathBuf,
+}
```

Comment:
Heh, this is tricky.... `ide` layer absolutely shouldn't know about this information. It should not assume cargo or any other build system (or existence of a build system at all). 

The information about the root should be threaded from the upper layer, `rust-analyzer`. This is what happens for runnables, for example. We kinda-sorta do this here as well, by accepting `Workspace` as an argument to `hover`, but it doesn't feel like this happens on the right level of abstraction. 

I think that, instead of adding more parameters to hover, we should have conceptually conceptually this API

```rust
pub fn hover(&self, position: FilePosition) -> Markup { ... }

impl Markaup {
    fn as_plain_text(&self) -> String {}
    fn as_markdown(&self, config: MarkdownConfig) -> String {}
}

struct MarkdownConfig {
  links: bool,
  base_url: Option<String>,
}
```

Note that `MarkdownConfig` doesn't talk about target dir, workspaces, or any other build-system-specifc functionality

I am not sure how `Markup` should be represented internally. I guess, any of three reprs would work:

```rust
// Do all processing lazily
struct Markup { raw_markdown: String } 

// Lightly pre-processed to extract links
struct Markup {
  raw_markdown: Stirng,
  links: Vec<TextRange>, 
}

// Fully parsed markdown dom
type Markup = some_md_lib::Markdown
```

The first design is probably the best one.